src/complex_fluids/ComplexFluidForcing.cpp

```diff
@@ -0,0 +1,957 @@
+        visit_data_writer->registerPlotQuantity("Conformation_Tensor", "TENSOR", d_conform_idx_draw);
+    }
+    d_stress_draw = input_db->getBoolWithDefault("output_stress_tensor", d_stress_draw);
+    if (d_stress_draw)
+    {
+        d_stress_var_draw = new CellVariable<NDIM, double>(d_object_name + "::stress_draw", NDIM * NDIM);
+        d_stress_idx_draw = var_db->registerVariableAndContext(d_stress_var_draw, d_context, IntVector<NDIM>(0));
+        visit_data_writer->registerPlotQuantity("Stress_Tensor", "TENSOR", d_stress_idx_draw);
+    }
+
+    // Set up AMR tagging if necessary
+    d_divW_draw = input_db->getBoolWithDefault("output_divergence", d_divW_draw);
+    d_divW_rel_tag = input_db->getBoolWithDefault("divergence_rel_tagging", d_divW_rel_tag);
+    d_divW_abs_tag = input_db->getBoolWithDefault("divergence_abs_tagging", d_divW_abs_tag);
+    if (d_log_divW || d_divW_draw || d_divW_abs_tag || d_divW_rel_tag)
+    {
+        d_divW_var_draw = new CellVariable<NDIM, double>(d_object_name + "::divW_draw", NDIM);
+        d_divW_idx_draw = var_db->registerVariableAndContext(d_divW_var_draw, d_context, IntVector<NDIM>(0));
+        if (d_divW_draw) visit_data_writer->registerPlotQuantity("Stress_Divergence", "VECTOR", d_divW_idx_draw);
+    }
+    if (d_divW_rel_tag) d_divW_rel_thresh = input_db->getDoubleArray("divergence_rel_thresh");
+    if (d_divW_abs_tag) d_divW_abs_thresh = input_db->getDoubleArray("divergence_abs_thresh");
+    if (d_divW_abs_tag || d_divW_rel_tag)
+    {
+        d_adv_diff_integrator->registerApplyGradientDetectorCallback(&applyGradientDetectorCallback, this);
+    }
+
+    // Create boundary conditions for advected materials if not periodic
+    const IntVector<NDIM>& periodic_shift = grid_geom->getPeriodicShift();
+    if (periodic_shift.min() <= 0)
+    {
+        vector<RobinBcCoefStrategy<NDIM>*> conc_bc_coefs(NDIM * (NDIM + 1) / 2);
+        d_conc_bc_coefs.resize(NDIM * (NDIM + 1) / 2);
+        for (int d = 0; d < NDIM * (NDIM + 1) / 2; ++d)
+        {
+            ostringstream conc_bc_name;
+            conc_bc_name << "c_bc_coef_" << d;
+            const string bc_coefs_name = conc_bc_name.str();
+
+            ostringstream bc_coefs_db_name_stream;
+            bc_coefs_db_name_stream << "ExtraStressBoundaryConditions_" << d;
+            const string bc_coefs_db_name = bc_coefs_db_name_stream.str();
+
+            d_conc_bc_coefs[d] =
+                new muParserRobinBcCoefs(bc_coefs_name, input_db->getDatabase(bc_coefs_db_name), grid_geom);
+        }
+        d_adv_diff_integrator->setPhysicalBcCoefs(d_W_cc_var, d_conc_bc_coefs);
+    }
+    d_convec_oper = new AdvDiffComplexFluidConvectiveOperator(
+        "ComplexFluidConvectiveOperator", d_W_cc_var, input_db, d_convec_oper_type, d_conc_bc_coefs, vel_bcs);
+    d_adv_diff_integrator->setConvectiveOperator(d_W_cc_var, d_convec_oper);
+
+    // Register relaxation function
+    if (d_fluid_model == "OLDROYDB")
```

Comment:
Would be nice to make this more extensible (e.g. to allow for user specified models).